### PR TITLE
[New source] MangaInUa

### PR DESCRIPTION
**Pull request by ztimms73**
Ukrainian source request #14
TODO:
- [x] Search
- [x] Details
- [x] Pages

### CODE REVIEW

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+		query: String?,
+		tags: Set<MangaTag>?,
+		sortOrder: SortOrder?
+	): List<Manga> {
+		val page = (offset / 24f).toIntUp().inc()
+		val url = when {
+			!query.isNullOrEmpty() -> parseFailed("Search currently unavailable") // TODO
+			tags.isNullOrEmpty() -> "/mangas/page/$page".withDomain()
+			tags.size == 1 -> "${tags.first().key}/page/$page"
+			tags.size > 1 -> parseFailed("This source supports only 1 tag") // Maybe use string from resources?
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:30]:
```suggestion
			tags.size > 1 -> throw IllegalArgumentException("This source supports only 1 tag")
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+			else -> "/mangas/${page}".withDomain()
+		}
+		val doc = context.httpGet(url).parseHtml()
+		val container = doc.body().select("div.container") ?: parseFailed("Container not found")
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:34]:
```suggestion
		val container = doc.body().getElementById("dle-content") ?: parseFailed("Container not found")
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+				title = item.selectFirst("h3.card__title")?.text() ?: return@mapNotNull null,
+				coverUrl = item.selectFirst("header.card__cover")?.selectFirst("img")?.attrAsAbsoluteUrl("data-src").orEmpty(),
+				altTitle = null,
+				author = null,
+				rating = runCatching {
+					item.selectFirst("div.card__short-rate--num")
+						?.text()
+						?.toFloatOrNull()
+						?.div(10F)
+				}.getOrNull() ?: 0.5f,
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:49]:
```suggestion
				}.getOrNull() ?: RATING_UNKNOWN,
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+				url = href,
+				isNsfw = item.selectFirst("ul.card__list")?.select("li")?.last()?.text() == "18+",
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:51]:
```suggestion
				isNsfw = item.selectFirst("ul.card__list")?.select("li")?.lastOrNull()?.text() == "18+",
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+					item.selectFirst("div.card__category")?.select("a")?.mapToSet {
+						MangaTag(
+							title = it.ownText(),
+							key = it.attr("href").substringAfterLast('/').urlEncoded(),
+							source = source,
+						)
+					}
+				}.getOrNull().orEmpty(),
+				state = null,
+				publicUrl = href.withDomain(),
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:62]:
```suggestion
				publicUrl = href.toAbsoluteUrl(container.host ?: getDomain()),
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+	override suspend fun getTags(): Set<MangaTag> {
+		val domain = getDomain()
+		val doc = context.httpGet("https://$domain/mangas").parseHtml()
+		val root = doc.body().getElementById("menu_1")?.selectFirst("div.menu__wrapper") ?: parseFailed("Cannot find root")
+		return root.select("li").mapToSet { li ->
+			val a = li.selectFirst("a") ?: throw ParseException("a is null")
+			MangaTag(
+				title = a.ownText(),
+				key = a.attr("href"),
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:84]:
```suggestion
				key = a.attr("href").removeSuffix("/").substringAfterLast('/'),
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+						?.toFloatOrNull()
+						?.div(10F)
+				}.getOrNull() ?: 0.5f,
+				url = href,
+				isNsfw = item.selectFirst("ul.card__list")?.select("li")?.last()?.text() == "18+",
+				tags = runCatching {
+					item.selectFirst("div.card__category")?.select("a")?.mapToSet {
+						MangaTag(
+							title = it.ownText(),
+							key = it.attr("href").substringAfterLast('/').urlEncoded(),
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:56]:
```suggestion
							key = it.attr("href").removeSuffix("/").substringAfterLast('/'),
```

File: src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+import org.koitharu.kotatsu.parsers.config.ConfigKey
+import org.koitharu.kotatsu.parsers.exception.ParseException
+import org.koitharu.kotatsu.parsers.model.*
+import org.koitharu.kotatsu.parsers.util.*
+
+@MangaSourceParser("MANGAINUA", "MANGA/in/UA", "uk")
+class MangaInUaParser(override val context: MangaLoaderContext) : MangaParser(MangaSource.MANGAINUA) {
+
+	override val sortOrders: Set<SortOrder>
+		get() = emptySet() // Isn't supported
```
Comment (nv95) [inline src/main/kotlin/org/koitharu/kotatsu/parsers/site/MangaInUaParser.kt:15]:
```suggestion
		get() = Collections.singleton(SortOrder.UPDATED)
```

**Comment by ztimms73** [comment]
I think it is needed before merging to squash these 4 commits and then merge

**Pull request merged**